cosmwasm/packages/wasmi-runtime/src/wasm/query_chain.rs

```diff
@@ -0,0 +1,278 @@
+use super::errors::WasmEngineError;
+use crate::crypto::Ed25519PublicKey;
+use crate::wasm::types::{IoNonce, SecretMessage};
+use crate::{exports, imports};
+
+use crate::cosmwasm::encoding::Binary;
+use crate::cosmwasm::query::{QueryRequest, WasmQuery};
+use crate::cosmwasm::{
+    std_error::{StdError, StdResult},
+    system_error::{SystemError, SystemResult},
+};
+
+use enclave_ffi_types::{Ctx, EnclaveBuffer, OcallReturn, UntrustedVmError};
+use log::*;
+use sgx_types::sgx_status_t;
+
+pub fn encrypt_and_query_chain(
+    query: &[u8],
+    context: &Ctx,
+    nonce: IoNonce,
+    user_public_key: Ed25519PublicKey,
+) -> Result<(Vec<u8>, u64), WasmEngineError> {
+    let mut query_struct: QueryRequest = match serde_json::from_slice(query) {
+        Ok(query_struct) => query_struct,
+        Err(err) => {
+            error!(
+                "encrypt_and_query_chain() cannot build struct from json {:?}: {:?}",
+                String::from_utf8_lossy(query),
+                err
+            );
+            let answer: SystemResult<StdResult<Binary>> = Err(SystemError::InvalidRequest {
+                request: Binary(query.into()),
+                error: String::from(format!("{}", err)),
+            });
+
+            let answer_as_vec = serde_json::to_vec(&answer).map_err(|err| {
+                // this should never happen
+                error!(
+                    "encrypt_and_query_chain() got an error while trying to serialize the error {:?} returned to WASM: {:?}",
+                    answer,
+                    err
+                );
+
+                WasmEngineError::SerializationError
+            })?;
+
+            return Ok((answer_as_vec, 500)); // Should we charge gas for this to prevent spam?
+        }
```

Comment:
>Should we charge gas for this to prevent spam?

I think it doesn't really matter since the contract will still have to run some code inside the vm between consecutive calls to imports. That being said, this can't hurt since users shouldn't encounter it in regular use.